pkg/ioutil/ioutil.go

```diff
@@ -205,47 +205,55 @@ func CopyAligned(w *os.File, r io.Reader, alignedBuf []byte) (int64, error) {
 	}
 
 	var written int64
-	var err error
 	for {
-		nr, er := r.Read(alignedBuf)
-		if nr == len(alignedBuf) {
-			// Buffer read is aligned with input buffer, proceed to write.
-			nw, ew := w.Write(alignedBuf)
+		buf := alignedBuf
+		if totalSize != -1 {
+			remaining := totalSize - written
+			if remaining < int64(len(buf)) {
+				buf = buf[:remaining]
+			}
+		}
+		nr, err := io.ReadFull(r, buf)
+		if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
+			return written, err
+		}
+		eof := err == io.EOF || err == io.ErrUnexpectedEOF
+		buf = buf[:nr]
+		if len(buf)%directioAlignSize == 0 {
+			// buf is aligned for directio write()
+			var nw int
+			nw, err = w.Write(buf)
 			if nw > 0 {
 				written += int64(nw)
 			}
-			if ew != nil {
-				err = ew
-				break
+			if err != nil {
+				return written, err
 			}
 			if nr != nw {
-				err = io.ErrShortWrite
-				break
+				return written, io.ErrShortWrite
 			}
-		} else if nr > 0 {
-			// Buffer read is not aligned with input buffer, proceed to write
-			// whatever possible as aligned and turn off direct I/O.
-			nw, ew := writeUnaligned(w, alignedBuf[:nr])
+		} else {
+			// buf is not aligned, hence use writeUnaligned()
+			var nw int
+			nw, err = writeUnaligned(w, buf[:nr])
 			if nw > 0 {
 				written += int64(nw)
 			}
-			if ew != nil {
-				err = ew
-				break
+			if err != nil {
+				return written, err
 			}
 			if nr != nw {
-				err = io.ErrShortWrite
-				break
+				return written, io.ErrShortWrite
```

Comment:
this if else block is repeating a bit - suggest refactoring it to
```
var nw int
if len(buf)%directioAlignSize == 0 {
     // buf is aligned for directio write()
      nw, err = w.Write(buf)
else {
      nw, err = writeUnaligned(w, buf[:nr])
}
if nw > 0 {
	written += int64(nw)
}
if err != nil {
	return written, err
}
if nr != nw {
	return written, io.ErrShortWrite
}

Reply:
Can this incorporated @krishnasrinivas ?